ott/core/initializers.py

```diff
@@ -278,6 +282,252 @@ def init_dual_a(
+  ):
+    self.geom = geom
+    self.dtype = geom.x.dtype
+    self.opt = opt
+
+    na, nb = geom.shape
+    if meta_model is None:
+      # Default to a Meta_MLP if not specified.
+      self.meta_model = Meta_MLP(
+          num_hidden_units=512,
+          num_hidden_layers=2,
+          potential_size=na,
+      )
+    else:
+      self.meta_model = meta_model
+
+    # Initialize the model.
+    a_placeholder = jnp.zeros(na, dtype=self.dtype)
+    b_placeholder = jnp.zeros(nb, dtype=self.dtype)
+    params = self.meta_model.init(init_key, a_placeholder,
+                                  b_placeholder)['params']
+    self.state = train_state.TrainState.create(
+        apply_fn=self.meta_model.apply, params=params, tx=opt
+    )
+    self.update_impl = jax.jit(self._get_update_fn())
+
+  def update(self, state: train_state.TrainState, a: jnp.array, b: jnp.array):
+    r"""Update the meta model with the dual objective.
+
+    The goal is for the model to match the optimal duals, i.e.,
+    :math:`\hat f_\theta \approx f^\star`.
+    This can be done by training the predictions of :math:`\hat f_\theta`
+    to optimize the dual objective, which :math:`f^\star` also optimizes for.
+    The overall learning setup can thus be written as:
+
+    .. math::
+      \min_\theta\; {\mathbb E}_{(\alpha,\beta)\sim{\mathcal{D}}}\; J(\hat f_\theta(a, b); \alpha, \beta),
+
+
+    where :math:`a,b` are the probabilities of the measures :math:`\alpha,\beta`,
+    :math:`\mathcal{D}` is a meta distribution of optimal transport problems,
+
+    .. math::
+      -J(f; \alpha, \beta, c) := \langle f, a\rangle + \langle g, b \rangle - \epsilon\left\langle \exp\{f/\epsilon\}, K\exp\{g/\epsilon\}\right\rangle
+
+    is the entropic dual objective,
+    and :math:`K_{i,j} := -C_{i,j}/\epsilon` is the *Gibbs kernel*.
+
+    Args:
+      state: Optimizer state of the meta model.
+      a: Probabilites of the :math:`\alpha` measure's atoms
+      b: Probabilites of the :math:`\beta` measure's atoms
+
+    Returns:
+      The training loss, :math:`f`, and updated state
+    """
+    return self.update_impl(state, a, b)
+
+  def init_dual_a(
+      self, ot_prob: linear_problems.LinearProblem, lse_mode: bool
+  ) -> jnp.ndarray:
+    # Detect if the problem is batched.
+    if ot_prob.a.ndim == 1:
+      vmap_a_val = None
+    else:
+      assert ot_prob.a.ndim == 2
+      vmap_a_val = 0
+
+    if ot_prob.b.ndim == 1:
+      vmap_b_val = None
+    else:
+      assert ot_prob.b.ndim == 2
+      vmap_b_val = 0
+
+    if vmap_a_val is not None or vmap_b_val is not None:
+      compute_f_maybe_batch = jax.vmap(
+          self._compute_f, in_axes=(vmap_a_val, vmap_b_val, None)
+      )
+    else:
+      compute_f_maybe_batch = self._compute_f
+
+    init_f = compute_f_maybe_batch(ot_prob.a, ot_prob.b, self.state.params)
+    f_u = init_f if lse_mode else ot_prob.geom.scaling_from_potential(init_f)
+    return f_u
+
+  def _dual_obj_from_f(self, a, b, f):
```

Comment:
if feels (apart from update to g) that this is doing something similar [to `ent_reg_cost` in Sinkhorn](https://github.com/ott-jax/ott/blob/468c2d54686699a23547cfa15c608a536f838fe7/ott/core/sinkhorn.py#L143), could we consider reusing that part of the code?

Reply:
Yeah, I just updated to using `ent_reg_cost` in https://github.com/ott-jax/ott/pull/145/commits/7ab02a1f1640f28b9c7fad2c30dafddd68e60151. I originally was hesitant to instantiate the `LinearProblem` to call this but it doesn't seem like it introduces much overhead.